worldedit-core/src/main/java/com/sk89q/worldedit/extension/factory/parser/DefaultBlockParser.java

```diff
@@ -272,21 +280,41 @@ private BaseBlock parseLogic(String input, ParserContext context) throws InputPa
 
         if (state == null) {
             String typeString;
-            String stateString = null;
+
             int stateStart = blockAndExtraData[0].indexOf('[');
-            if (stateStart == -1) {
+            int nbtStart = blockAndExtraData[0].indexOf('{');
+            int typeEnd = stateStart == -1 ? nbtStart : stateStart;
+
+            if (typeEnd == -1) {
                 typeString = blockAndExtraData[0];
             } else {
-                typeString = blockAndExtraData[0].substring(0, stateStart);
+                typeString = blockAndExtraData[0].substring(0, typeEnd);
+            }
+
+            String stateString = null;
+            if (stateStart != -1) {
                 if (stateStart + 1 >= blockAndExtraData[0].length()) {
                     throw new InputParseException(TranslatableComponent.of("worldedit.error.parser.hanging-lbracket", TextComponent.of(stateStart)));
                 }
                 int stateEnd = blockAndExtraData[0].lastIndexOf(']');
```

Comment:
Check behavior for this test case: `chest[double={whoops:"nbt ]"}`; it might pass, or fail with a weird error.

I suspect we might need to be a little more strict in what is accepted here. It might be worth writing an ANTLR grammar for this instead, since it's getting quite complex.

Reply:
`State is missing trailing ']'` is the error we get here, which appears to be the correct error